### DISCUSSION TITLE
Downcast subtype operator

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#7991] [Language Ideas] [answered]
The cast operator using () or “as” allows writing any target type.
It would be nicer to have a restriction only for subtypes.
It is possible to write such extension method now but only a specific type. 
Generics do not support high order types to write a single method for any type.

Before:
```cs
MyType m:

(String)m // compiles , always fail in runtime
(MySubType) // compiles
```

workaround:
```cs
class MyTypeExt{
 public static T Cast<T>(this MyType m) where T:MyType => (T)m;
}

// repeat for any type needed
```

With subtype operator: (syntax just an example)

```cs
m subtype_cast string // doesn’t compile
m subtype_cast MySubType // compiles
```

**Comment by KennethHoff** [comment, marked as answer]
By default both `x as X` and `(X)x` gives you a warning if it's not a hierarchy. Make those warnings into errors and you're good - no changes required.

**Comment by HaloFour** [comment]
> The cast operator using () or “as” allows writing any target type.

Are you sure?  If I try to cast from `MyType` to `string` I already get error CS0030 or CS0039.  I have to upcast to `object` first before it will let me downcast to `string`.
